include/nbl/asset/ICPUDescriptorSet.h

```diff
@@ -144,57 +171,37 @@ class NBL_API ICPUDescriptorSet final : public IDescriptorSet<ICPUDescriptorSetL
-				return m_bindingInfo->operator[](index).descriptorType;
-			return EDT_INVALID;
-		}
+			const uint32_t descriptorOffset = getLayout()->getDescriptorOffset(bindingInfo->type, binding);
+			const uint32_t descriptorCount = bindingInfo->count;
 
-		//! Can modify the array of descriptors bound to a particular bindings
-		inline core::SRange<SDescriptorInfo> getDescriptors(uint32_t index) 
-		{ 
-			assert(!isImmutable_debug());
+			auto descriptorsBegin = m_descriptors[bindingInfo->type]->begin() + descriptorOffset;
+			auto descriptorInfosBegin = m_descriptorInfos[bindingInfo->type]->begin() + descriptorOffset;
 
-			if (m_bindingInfo && index<m_bindingInfo->size())
-			{
-				const auto& info = m_bindingInfo->operator[](index);
-				auto _begin = m_descriptors->begin()+info.offset;
-				if (index+1u!=m_bindingInfo->size())
-					return core::SRange<SDescriptorInfo>{_begin, m_descriptors->begin()+m_bindingInfo->operator[](index+1u).offset};
-				else
-					return core::SRange<SDescriptorInfo>{_begin, m_descriptors->end()};
-			}
-			else
-				return core::SRange<SDescriptorInfo>{nullptr, nullptr};
-		}
-		inline core::SRange<const SDescriptorInfo> getDescriptors(uint32_t index) const
-		{
-			if (m_bindingInfo && index<m_bindingInfo->size())
-			{
-				const auto& info = m_bindingInfo->operator[](index);
-				auto _begin = m_descriptors->begin()+info.offset;
-				if (index+1u!=m_bindingInfo->size())
-					return core::SRange<const SDescriptorInfo>{_begin, m_descriptors->begin()+m_bindingInfo->operator[](index+1u).offset};
-				else
-					return core::SRange<const SDescriptorInfo>{_begin, m_descriptors->end()};
-			}
-			else
-				return core::SRange<const SDescriptorInfo>{nullptr, nullptr};
+			return { {descriptorsBegin, descriptorsBegin+descriptorCount}, {descriptorInfosBegin, descriptorInfosBegin+descriptorCount} };
 		}
 
-		inline auto getTotalDescriptorCount() const
+		core::SRange<core::smart_refctd_ptr<ICPUSampler>> getMutableSamplers(const uint32_t binding) const
 		{
-			return m_descriptors->size();
+			const uint32_t offset = getLayout()->getMutableSamplerOffset(binding);
+			if (offset == ~0u)
+				return { nullptr, nullptr };
+
+			const auto bindingInfo = std::lower_bound(getLayout()->getBindings().begin(), getLayout()->getBindings().end(), ICPUDescriptorSetLayout::SBinding{ binding });
+			assert(bindingInfo->binding == binding && "binding is not in the descriptor set!");
+
+			auto samplersBegin = m_mutableSamplers->begin() + offset;
+			return { samplersBegin, samplersBegin + bindingInfo->count };
 		}
 
+		inline core::smart_refctd_ptr<IDescriptor>* getDescriptorStorage(const E_DESCRIPTOR_TYPE type) const { return m_descriptors[type]->begin(); }
+		inline SDescriptorInfo::SBufferImageInfo* getDescriptorInfoStorage(const E_DESCRIPTOR_TYPE type) const { return m_descriptorInfos[type]->begin(); }
+		inline core::smart_refctd_ptr<ICPUSampler>* getMutableSamplerStorage() const { return m_mutableSamplers->begin(); }
```

Comment:
1. feels like we should really just keep `m_descriptorInfos` and get rid of `m_descriptors` since there's duplicate data
2. these methods should not have `const` on the end, as they return non-const pointers
3. but they should ahve `non-const` variants that do the mutability checks like other Asset methods
4. maybe consider adding overloads with `const E_DESCRIPTOR_TYPE, const binding_t` which can do the searches using `IDescriptorSetLayout::SBindingRedirect`'s `operator[]` and give the right pointer right away

Reply:
> these methods should not have const on the end, as they return non-const pointers

[Their usage in the asset converter mandates the `const`.](https://github.com/Devsh-Graphics-Programming/Nabla/blob/22284fd82eb56e2dcec09df4f486d620bb82b474/include/nbl/video/utilities/IGPUObjectFromAssetConverter.h#L1640-L1681)

Reply:
> but they should ahve non-const variants that do the mutability checks like other Asset methods

> maybe consider adding overloads with const E_DESCRIPTOR_TYPE, const binding_t which can do the searches using IDescriptorSetLayout::SBindingRedirect's operator[] and return an core::SRange<>

Don't think they are required any where right now.

Reply:
> feels like we should really just keep m_descriptorInfos and get rid of m_descriptors since there's duplicate data

Resolved.

Reply:
> MAybe even the regular non-binding specific should return an SRange ?

I see no point, but sure, if you want.

Reply:
@achalpandeyy put the mutability checks there, comment them out and put a link to these convos in the comments

Reply:
@Hazardu this is the reason for the mapping of Asset to CPU Object + Metadata so we can patch up the usage flags and shit while keeping the Assets const

Reply:
> @achalpandeyy put the mutability checks there, comment them out and put a link to these convos in the comments

I think these comments are old because now I already have mutability checks in [there](https://github.com/Devsh-Graphics-Programming/Nabla/blob/63bea089173daa1cc0cf6a75cc4ab9fb461be5fc/src/nbl/asset/ICPUDescriptorSet.cpp#L8).

Reply:
Resolved.